rust-src/wallet_library/src/statement.rs

```diff
@@ -0,0 +1,375 @@
+use concordium_base::{
+    base::ContractAddress,
+    common::Serialize,
+    curve_arithmetic::Curve,
+    id::{id_proof_types::AtomicStatement, types::*},
+    web3id::{CredentialStatement, Request},
+};
+use std::{
+    collections::{BTreeMap, HashSet},
+    fmt::Display,
+    hash::Hash,
+    marker::PhantomData,
+};
+
+pub trait AcceptableRequest<C: Curve, AttributeType: Attribute<C::Scalar>> {
+    /// Method to check whether the request is acceptable given the provided
+    /// config. i.e. that it satifies the rules in the config, and some
+    /// general rules.
+    fn acceptable_request(
+        &self,
+        config: &WalletConfig<C, AttributeType>,
+    ) -> AcceptableRequestResult;
+}
+
+pub trait AcceptableAtomicStatement<
+    C: Curve,
+    TagType: AttributeTagType,
+    AttributeType: Attribute<C::Scalar>,
+> {
+    /// Method to check that the statement is acceptable given the provided
+    /// rules.
+    fn acceptable_atomic_statement(
+        &self,
+        rules: &Option<WalletConfigRules<C, TagType, AttributeType>>,
+    ) -> AcceptableRequestResult;
+}
+
+pub type AcceptableRequestResult = Result<(), RequestCheckError>;
+
+pub struct WalletConfig<'a, C: Curve, AttributeType: Attribute<C::Scalar>> {
+    /// Rules that statements on identity credentials should satisfy
+    pub identity_rules: Option<WalletConfigRules<'a, C, AttributeTag, AttributeType>>,
+    /// Rules that statements on web3Id credentials should satisfy. If a
+    /// statement uses a contract index without an entry, it should only
+    /// consider for basic, global, rules.
+    pub web3_rules:     BTreeMap<ContractAddress, WalletConfigRules<'a, C, String, AttributeType>>,
+}
+
+pub struct WalletConfigRules<
+    'a,
+    C: Curve,
+    TagType: AttributeTagType,
+    AttributeType: Attribute<C::Scalar>,
+> {
+    /// The set of tags, which are allowed to be used for range statements.
+    pub range_tags:      HashSet<TagType>,
+    /// The set of tags, which are allowed to be used for membership and
+    /// nonMembership statements.
+    pub set_tags:        HashSet<TagType>,
+    /// A function to check attributes using custom behaviour.
+    pub attribute_check: AttributeCheck<'a, TagType, AttributeType>,
+    pub _marker:         PhantomData<C>,
+}
+
+pub trait AttributeTagType: Serialize + Ord + Hash + Display {}
+impl<T: Serialize + Ord + Hash + Display> AttributeTagType for T {}
+
+type AttributeCheck<'a, TagType, AttributeType> =
+    Box<dyn Fn(&TagType, &AttributeType) -> AcceptableRequestResult + 'a>;
```

Comment:
Why this complicated lifetime when it always seems to be `'static` at the end?

Reply:
For the default config it is just static, but when others use it (and add web3Id rules) I think it's better to have this. :thinking:

Reply:
Ok. I'm generally of the opinion that complexity should be there if it's needed, and abstractions should be introduced when they are needed, not when they are thought that they might be needed.

But OK :)